torchgeo/datasets/ssl4eo.py

```diff
@@ -0,0 +1,229 @@
+       takes roughly 36 hrs to download, 1 hr to checksum, and XX hrs to extract.
+
+    .. versionadded:: 0.5
+    """  # noqa: E501
+
+    locations = 251079
+    times = 4
+    size = 264
+
+    metadata = {
+        "s1": {
+            "filename": "s1.tar.gz",
+            "md5": "51ee23b33eb0a2f920bda25225072f3a",
+            "bands": ["VV", "VH"],
+        },
+        "s2c": {
+            "filename": "s2_l1c.tar.gz",
+            "md5": "b4f8b03c365e4a85780ded600b7497ab",
+            "bands": [
+                "B1",
+                "B2",
+                "B3",
+                "B4",
+                "B5",
+                "B6",
+                "B7",
+                "B8",
+                "B8A",
+                "B9",
+                "B11",
+                "B12",
+            ],
+        },
+        "s2a": {
+            "filename": "s2_l2a.tar.gz",
+            "md5": "85496cd9d6742aee03b6a1c99cee0ac1",
+            "bands": [
+                "B1",
+                "B2",
+                "B3",
+                "B4",
+                "B5",
+                "B6",
+                "B7",
+                "B8",
+                "B8A",
+                "B9",
+                "B11",
+                "B12",
+            ],
+        },
+    }
+
+    def __init__(
+        self,
+        root: str = "data",
+        split: str = "s2c",
+        transforms: Optional[Callable[[Dict[str, Tensor]], Dict[str, Tensor]]] = None,
+        checksum: bool = False,
+    ) -> None:
+        """Initialize a new SeCo dataset instance.
+
+        Args:
+            root: root directory where dataset can be found
+            split: one of "s1" (Sentinel-1 dual-pol SAR), "s2c" (Sentinel-2 Level-1C
+                top-of-atmosphere reflectance), and "s2a" (Sentinel-2 Level-2a surface
+                reflectance
+            transforms: a function/transform that takes input sample and its target as
+                entry and returns a transformed version
+            checksum: if True, check the MD5 of the downloaded files (may be slow)
+
+        Raises:
+            AssertionError: if ``split`` argument is invalid
+            RuntimeError: if dataset is missing or checksum fails
+        """
+        assert split in self.metadata
+
+        self.root = root
+        self.split = split
+        self.transforms = transforms
+        self.checksum = checksum
+
+        self._verify()
+
+    def __getitem__(self, index: int) -> Dict[str, Tensor]:
+        """Return an index within the dataset.
+
+        Args:
+            index: index to return
+
+        Returns:
+            sample with an "image" in 5xCxHxW format where the 5 indexes over the same
+                patch sampled from different points in time by the SeCo method
+        """
+        directory = os.path.join(self.root, self.split, f"{index // self.times:07}")
+        subdirs = sorted(os.listdir(directory))
+        directory = os.path.join(directory, subdirs[index % self.times])
```

Comment:
This is somewhat hard-coded, but is much faster than storing a list of every directory like we do in SeCo.